### ISSUE TITLE
opam-dune-lint gets the wrong package for some dependencies

### ISSUE THREAD

**Issue by talex5** [ocurrent/opam-dune-lint#53]
e.g. with Eio, ocaml-ci is currently reporting:

```
eio_main.opam: changes needed:
  "fmt" {>= "0.9.0"}                       [from lib_eio_linux, lib_eio_posix, lib_eio_windows]
  "iomux" {>= "0"}                         [from lib_eio_posix]
  "logs" {>= "0.7.0"}                      [from lib_eio_linux]
  "uring" {>= "0"}                         [from lib_eio_linux]
```
But none of these directories is (directly) part of the eio_main package anyway.

For example, `iomux` is a dependency only of the `eio_posix` library, which is part of the `eio_posix` package (see https://github.com/ocaml-multicore/eio/blob/60e67472e3726192c4f6b9fbff81db5eb15b8a27/lib_eio_posix/dune#L10).

If I add `(package eio_posix)` to this, dune complains:
```
Error: This library has a public_name, it already belongs to the package eio_posix
```
And `dune describe` says:
```
  (library
   ((names (eio_posix))
    (extensions ())
    (package (eio_posix))
    (source_dir lib_eio_posix)
    (external_deps
     ((fmt required)
      (iomux required)))
    (internal_deps
     ((eio required)
      (eio.utils required)
      (eio.unix required)))))
```
So dune is reporting that `iomux` is part of `eio_posix`.

**Comment by moyodiallo** [comment]
This is a bug, by default opam-dune-lint also resolve the transitive dependencies because a public library could depends on a private library. It doesn't resolve the internal_deps which are public.

`dune describe` doesn't printing `lib_eio_linux`, `lib_eio_posix`, `lib_eio_windows` as  internal_deps, because of the alternative `select` dune stanza.
```
(library
  (name eio_main)
  (public_name eio_main)
  (libraries
    (select linux_backend.ml from
            (eio_linux -> linux_backend.enabled.ml)
            (          -> linux_backend.disabled.ml))
    (select posix_backend.ml from
            (eio_posix -> posix_backend.enabled.ml)
            (          -> posix_backend.disabled.ml))
    (select windows_backend.ml from
            (eio_windows -> windows_backend.enabled.ml)
            (            -> windows_backend.disabled.ml))
    ))
```
This is why we're seeing this bug. It could be easily solved  by sharing a list of all the public libraries.

**Comment by moyodiallo** [comment]
> ~~dune describe doesn't printing lib_eio_linux, lib_eio_posix, lib_eio_windows as internal_deps, because of the alternative select dune stanza.~~

the command `dune describe external-lib-deps` gives:
 ```
(library
   ((names (eio_main))
    (extensions ())
    (package (eio_main))
    (source_dir lib_main)
    (external_deps ())
    (internal_deps
     ((eio_linux optional)
      (eio_posix optional)
      (eio_windows optional)))))
```
My bad, my description above had some mistake. The internal optional public libs were resolve this is why the bug. Fixed here #54.

**Comment by moyodiallo** [comment]
The fix is deployed on ocaml-ci, is it OK to close this issue ? @talex5

**Comment by talex5** [comment]
That seems to work - thanks for the quick fix!

(it's now saying `WARNING: can't find opam package providing "runtime_events"`, but that's another problem!)
